src/properties/degrees.c

```diff
@@ -517,3 +517,138 @@ igraph_error_t igraph_sort_vertex_ids_by_degree(const igraph_t *graph,
+                                    const igraph_vector_int_t* weights) {
+    igraph_eit_t eit;
+    igraph_es_t es;
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t eid;
+    igraph_integer_t v1id;
+    igraph_integer_t v2id;
+    igraph_integer_t v1deg;
+    igraph_integer_t v2deg;
+
+    if (weights && igraph_vector_int_size(weights) != no_of_edges) {
+        IGRAPH_ERRORF("Weight vector length (%" IGRAPH_PRId ") does not match number of edges (%" IGRAPH_PRId ").",
+                      IGRAPH_EINVAL,
+                      igraph_vector_int_size(weights), no_of_edges);
+    }
+
+    if (igraph_is_directed(graph)) {
+        igraph_vector_int_t out_degrees;
+        igraph_vector_int_t in_degrees;
+        igraph_integer_t max_out_degree;
+        igraph_integer_t max_in_degree;
+
+        IGRAPH_CHECK(igraph_vector_int_init(&out_degrees, 0));
+        IGRAPH_CHECK(igraph_vector_int_init(&in_degrees, 0));
+        IGRAPH_CHECK(igraph_degree(graph, &out_degrees, igraph_vss_all(), IGRAPH_OUT, true));
+        IGRAPH_CHECK(igraph_degree(graph, &in_degrees, igraph_vss_all(), IGRAPH_IN, true));
+        IGRAPH_CHECK(igraph_maxdegree(graph, &max_out_degree, igraph_vss_all(), IGRAPH_OUT, true));
+        IGRAPH_CHECK(igraph_maxdegree(graph, &max_in_degree, igraph_vss_all(), IGRAPH_IN, true));
+
+        if (dout > 0 && din > 0) {
+            if (dout < max_out_degree || din < max_in_degree) {
+                IGRAPH_ERRORF("dout (%" IGRAPH_PRId ") or din (%" IGRAPH_PRId ") are smaller than the smallest required dimensions of the matrix (%" IGRAPH_PRId ", %" IGRAPH_PRId ")",
+                              IGRAPH_EINVAL,
+                              dout, din, max_out_degree, max_in_degree);
+            }
+            IGRAPH_CHECK(igraph_matrix_int_resize(jdm, dout, din));
+        } else {
+            IGRAPH_CHECK(igraph_matrix_int_resize(jdm, max_out_degree, max_in_degree));
+        }
+        // Set all elements to 0
+        igraph_matrix_int_null(jdm);
+
+        IGRAPH_CHECK(igraph_es_all(&es, IGRAPH_EDGEORDER_ID));
+        IGRAPH_CHECK(igraph_eit_create(graph, es, &eit));
+
+        while (!IGRAPH_EIT_END(eit)) {
+            eid = IGRAPH_EIT_GET(eit);
+            v1id = IGRAPH_FROM(graph, eid);
+            v2id = IGRAPH_TO(graph, eid);
+            v1deg = igraph_vector_int_get(&out_degrees,v1id);
+            v2deg = igraph_vector_int_get(&in_degrees, v2id);
+            if (!weights) {
+                MATRIX(*jdm, v1deg-1, v2deg-1)++;
+            } else {
+                MATRIX(*jdm, v1deg-1, v2deg-1) += VECTOR(*weights)[eid];
+            }
+            IGRAPH_EIT_NEXT(eit);
+        }
+
+        igraph_vector_int_destroy(&out_degrees);
+        igraph_vector_int_destroy(&in_degrees);
+
+    } else {
+        igraph_vector_int_t degrees;
+        igraph_integer_t max_degree;
+        // Compute max degrees
+        IGRAPH_CHECK(igraph_vector_int_init(&degrees, 0));
```

Comment:
You probably noticed now that some of the CI tests fail because of memory leaks. For example:

https://dev.azure.com/igraph-team/igraph/_build/results?buildId=5789&view=logs&j=1a52b443-5fb8-53f1-821f-633f5d086e98&t=16950513-9d69-54bb-da2c-179f24a13cca&l=1108

This happens not because of incorrect code, but because we did not yet add memory management. Looking at the error, 

```
Direct leak of 20 byte(s) in 1 object(s) allocated from:
    #0 0x7f03d84b4c38 in __interceptor_realloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:164
    #1 0x561732898b3b in igraph_vector_int_reserve /home/vsts/work/1/s/src/core/vector.pmt:474
    #2 0x56173289abc2 in igraph_vector_int_resize /home/vsts/work/1/s/src/core/vector.pmt:1270
    #3 0x5617328e0a9f in igraph_degree /home/vsts/work/1/s/src/graph/type_indexededgelist.c:1232
    #4 0x5617328e9c77 in igraph_construct_jdm /home/vsts/work/1/s/src/properties/degrees.c:573
    #5 0x56173286e866 in main /home/vsts/work/1/s/tests/unit/jdm.c:240
    #6 0x7f03d7a29d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
```

it tells us the line number to look at, line 240 in the test file. This is where you tested incorrect input. What happens is that when the function quits due to the `IGRAPH_ERROR` below, the `degrees` vector is not de-allocated. To fix this, we can do this:

```suggestion
        IGRAPH_CHECK(igraph_vector_int_init(&degrees, 0));
        IGRAPH_FINALLY(igraph_vector_int_destroy, &degrees);
```

This registers the vector for de-allocation in case of errors. We will also need to de-register it after we destroy it manually. Look for a comment on this below.

There is a shorthand for the above two lines,

```suggestion
        IGRAPH_VECTOR_INT_INIT_FINALLY(&degrees, 0);
```

This macro is here: https://github.com/igraph/igraph/blob/master/include/igraph_vector.h#L123

There is one more small improvement we can make. We can directly allocate the required vector size, so the `igraph_degree()` function wouldn't have to resize it later.

```suggestion
        IGRAPH_VECTOR_INT_INIT_FINALLY(&degrees, no_of_nodes);
```

This is a minor optimization which probably makes very little difference here, but I wanted to show you the principle. Let's go with this last version.

Please make this change for the directed version as well.